[Regression@296647@main] HDR canvas is no longer HDR

mwyrzykowski opened the pull request:
#### 57c0f48eb8d117e464e9da786f0591897882ec73
<pre>
[Regression@296647@main] HDR canvas is no longer HDR
<a href="https://bugs.webkit.org/show_bug.cgi?id=295301">https://bugs.webkit.org/show_bug.cgi?id=295301</a>
<a href="https://rdar.apple.com/154801239">rdar://154801239</a>

Reviewed by Gerald Squelart.

Only set contentsHeadroom on the CALayer when we disable tonemapping
otherwise setting CAToneMapMode to CAToneMapModeIfSupported will have
no effect.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):

Canonical link: <a href="https://commits.webkit.org/296899@main">https://commits.webkit.org/296899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8d0f4cef75c647ee2dd39a4ef1be010edf10fa5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60154 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38161 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83557 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98980 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63999 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23485 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59731 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93485 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17170 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118728 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36954 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27383 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92535 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95246 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92359 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23529 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37332 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15072 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32809 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36849 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42319 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36509 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38218 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -578,14 +578,15 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
-        else
+        else {
             layer.toneMapMode = CAToneMapModeNever;
+            layer.contentsHeadroom = m_maxRequestedEDRHeadroom;
```
This clash between contentsHeadroom and CAToneMapModeIfSupported doesn't seem to be documented (but I trust you've tested it). Should this be reported to the appropriate people, so they fix it or at least document it?

squelart commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -578,14 +578,15 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
     if (layer.wantsExtendedDynamicRangeContent) {
         // Painted contents have already been tonemapped, so disable it on the layer.
         if (isDelegatedDisplay)
             layer.toneMapMode = CAToneMapModeIfSupported;
```
Should you set layer.contentsHeadroom to 0 in this case? (Or will it never switch back&forth?)

mwyrzykowski replied:
oh yes, that's a good idea

squelart reviewed (approved):
r+ assuming you've tested it, and my comments don't require further action here.

mwyrzykowski committed:
[Regression@296647@main] HDR canvas is no longer HDR
https://bugs.webkit.org/show_bug.cgi?id=295301
rdar://154801239

Reviewed by Gerald Squelart.

Only set contentsHeadroom on the CALayer when we disable tonemapping
otherwise setting CAToneMapMode to CAToneMapModeIfSupported will have
no effect.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):

Canonical link: https://commits.webkit.org/296899@main
